### PR TITLE
CRUD Lesson and Lesson Word

### PR DESCRIPTION
**Pull request by abdullahallmehedi**

### CODE REVIEW

File: app/controllers/lessons_controller.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  def show
+  end
+  
+  def edit
+    @lesson_word = @lesson.lesson_words.unanswered.sample
+  end
+  
+  def update
+    @lesson_word = LessonWord.find params[:lesson][:lesson_word][:id]
```
Comment (chuanhtuan) [inline app/controllers/lessons_controller.rb]: in this feature, I want you using nested attribute
It means, you should display all words of answer on edit lesson screen and use javascript to hide/show word
and submit when finish all 20 words

File: app/assets/javascripts/lessons.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	
+	unhide();
+
+	editLessonForm.submit(function(event) {
+		console.log(index);
+		if (lessonWords.eq(index - 1).find('input:radio').filter(":checked").val()) {
+			if (index < 20) {
+				unhide();
+			} else {
+				console.log("bingo");
```
Comment (chuanhtuan) [inline app/assets/javascripts/lessons.js]: remove

File: app/controllers/lessons_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class LessonsController < ApplicationController
+  before_action :load_lesson, except: [:create]
+  
+  def create
+    @lesson = Lesson.new lesson_params
+    @lesson.user = current_user
+    words = @lesson.category.words.sample 20
```
Comment (chuanhtuan) [inline app/controllers/lessons_controller.rb]: @lesson.category.words.order("RANDOM()").limit(20)

File: app/models/lesson.rb [hunk 4]
```diff
@@ -1,6 +1,15 @@
...
-  has_many :lesson_words, dependent: :destroy
+  has_many :lesson_words, dependent: :destroy  
+  accepts_nested_attributes_for :lesson_words
   validates :result,  presence: true, numericality: { greater_than_or_equal_to: 0 }
-end
+  before_update :update_result
+  
+  def update_result
+    self.result = 0
+    lesson_words.each do |lesson_word|
```
Comment (chuanhtuan) [inline app/models/lesson.rb]: lesson_words.select do |lesson_word|
  lesson_word.word_answer.correct?
end.count

File: app/helpers/lessons_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module LessonsHelper
```
Comment (chuanhtuan) [inline app/helpers/lessons_helper.rb]: remove unused file

File: app/views/lessons/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<%= form_for @lesson, html: {id: "edit_lesson"} do |f| %>
+<div class="row">
+    <div class="col-xs-9 col-xs-offset-3" id="lesson_words">
+      <%= f.fields_for :lesson_words do |lwf| %>
+      <div class="row hidden">
+		  <div class="col-xs-4">
+		  	<%= lwf.object.word.content %>
+		  	<%= lwf.hidden_field :id, value: lwf.object.id %>
```
Comment (chuanhtuan) [inline app/views/lessons/edit.html.erb]: remove
no need

File: app/helpers/authentications_helper.rb [hunk 7]
```diff
@@ -32,5 +32,9 @@ def current_user? user
 
   def check_admin
     redirect_to categories_path unless current_user.admin?
-  end  
+  end
+  
+  def unanswered_lessons
```
Comment (chuanhtuan) [inline app/helpers/authentications_helper.rb]: where do you use this method

Comment (abdullahallmehedi) [inline app/helpers/authentications_helper.rb]: sry

File: .settings/org.eclipse.ltk.core.refactoring.prefs [hunk 8]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
Comment (linhnt) [inline .settings/org.eclipse.ltk.core.refactoring.prefs]: remove this file

File: app/assets/javascripts/lessons.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+$(document).ready(function() {
```
Comment (linhnt) [inline app/assets/javascripts/lessons.js]: re-indent tab with 2 whitespaces

File: app/controllers/lessons_controller.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
+class LessonsController < ApplicationController
+  before_action :load_lesson, except: [:create]
+  
+  def create
+    @lesson = Lesson.new lesson_params
+    @lesson.user = current_user
+    words = @lesson.category.words.order("RANDOM()").limit(20)
```
Comment (linhnt) [inline app/controllers/lessons_controller.rb]: dont use order random, get all lesson and use sample method

File: app/helpers/authentications_helper.rb [hunk 11]
```diff
@@ -32,5 +32,9 @@ def current_user? user
...
+    current_user.lessons.map {|lesson| lesson if lesson.lesson_words.unanswered.size > 0}
```
Comment (linhnt) [inline app/helpers/authentications_helper.rb]: current_user.lessons.select {|lesson| lesson.lesson_words.unanswered.size > 0}

File: app/models/lesson.rb [hunk 12]
```diff
@@ -1,6 +1,14 @@
...
   belongs_to :category
-  has_many :lesson_words, dependent: :destroy
+  has_many :lesson_words, dependent: :destroy  
+  accepts_nested_attributes_for :lesson_words
   validates :result,  presence: true, numericality: { greater_than_or_equal_to: 0 }
-end
+  before_update :update_result
+  
+  def update_result
+    self.result = lesson_words.select do |lesson_word|
```
Comment (linhnt) [inline app/models/lesson.rb]: remove self

File: app/controllers/lessons_controller.rb (hunk 10 above)
Comment (abdullahallmehedi) [inline app/controllers/lessons_controller.rb]: I have used sample function in my previous code. Then Tuan sir's comment was this:
@lesson.category.words.order("RANDOM()").limit(20)

What should I do know??

File: app/assets/javascripts/lessons.js (hunk 9 above)
Comment (abdullahallmehedi) [inline app/assets/javascripts/lessons.js]: I didn't understand your comment "re-indent tab with 2 whitespaces" on this line:
$(document).ready(function() {
This line have no tab. Though I have fixed other tab issues on that file.

Comment (linhnt) [inline app/assets/javascripts/lessons.js]: it's better now.

Comment (abdullahallmehedi) [inline app/assets/javascripts/lessons.js]: Thank You

**Pull request merged**